src/renderers/shared/reconciler/ReactReconciler.js

```diff
@@ -65,8 +65,10 @@ var ReactReconciler = {
     internalInstance, nextElement, transaction, context
   ) {
     var prevElement = internalInstance._currentElement;
-
-    if (nextElement === prevElement && nextElement._owner != null) {
+    if (nextElement === prevElement &&
+        nextElement._owner != null
+        // TODO: Shouldn't we need to do this: `&& context === internalInstance._context`
```

Comment:
@jimfb I tried your patch on my app which has huge precomputed static tree of React elements and without this line uncommented my components are not re-rendered on context changes.